### DISCUSSION TITLE
Save file backup system

### DISCUSSION THREAD

**Discussion by BradyThe** [XurxoMF/vs-launcher#22] [Suggestions]
One thing I really appreciate about the Stardew Valley launcher mod is the fact, that it back ups your save files. It helped me out more than once.

Currently I have a pretty convoluted system, that essentially copies the VS save file that has been modified last, before the bat starts the game. It'd be nice to have a more proper system, that wasn't hacked together with found batch commands, coded by somebody who knows what he is doing.
Could be a nice selling point for the mod too. How often I see comments asking for help because worlds have become corrupted.

I am no programmer, so I have no idea what the best way to go about this, would look like.
vcdbs get pretty large pretty fast, so a backup system probably shouldn't backup all save files indefinitely every time the launcher starts the game for two reasons: Storage capacity and backup speed.

Four thoughts:

- I think backing up worlds my way is a bit too restrictive.
- Keeping two backups per world/save file should suffice, deleting/overwriting the oldest.
- You could give the player the choice to backup selected save files in a general setting.
- You could give the player the choice to backup selected save files per Installation.

**Comment by XurxoMF** [comment]
Hi, thx for the suggestion.

A backup of the world and maybe some other data is something I already thought about a few days ago.

With the complete rewrite of the code I'm already implementing a config page and I'm changing downloads, extractions and deletions to asynchronous processes so I don't have to block the program for these things.

Once I have this I will add a configuration for each installation with which you can choose whether to make a backup or not and where.
By default they will be done in the folder of the installation but can be changed.
I will also add an option to configure how many backups to keep.

What I don't know is whether to do the backup before opening the game or when closing it. I'll have to see.

**Comment by trubbel2006** [comment]
> What I don't know is whether to do the backup before opening the game or when closing it. I'll have to see.

I reckon corruption of a save DB happens during game  or starting a game so taking a backup after closing seems counterproductive imho.
Maybe do a rolling backup before start and an optional rolling backup during game play (taking a snapshot every 15 minutes) ?
Optionally letting the user change the rollover amount and time between snapshots

**Reply by XurxoMF** [reply]
About making it before the game starts, it's a good point... making a corrupted backup it's useless.

About the backup during the game... I've to check how the game works. If Vintage Story keeps the world db open I can't make a backup... it'll throw errors.

I'll finish with the redesign and optimizations and I'll check that out.

**Comment by XurxoMF** [comment]
Backups Update is now available! Check it out on the releases page.
